scrapscript.py

```diff
@@ -4497,6 +4636,81 @@ def test_command(args: argparse.Namespace) -> None:
     unittest.main(argv=[__file__, *args.unittest_args])
 
 
+class ScrapError(Exception):
+    pass
+
+
+def _ensure_pygit2() -> ModuleType:
+    try:
+        import pygit2
+
+        assert isinstance(pygit2, ModuleType)
+        return pygit2
+    except ImportError:
+        raise ScrapError("Please install pygit2 to work with scrapyards")
+
+
+def yard_init(path: str) -> None:
+    git = _ensure_pygit2()
+    branch_name = "trunk"
+    repo = git.init_repository(path, initial_head=branch_name, bare=True)
+    try:
+        repo.head
+    except git.GitError:
+        pass
+    else:
+        raise ScrapError("Scrapyard already initialized; cannot init scrapyard again")
+    # Make an initial commit so that all other commands can do the normal
+    # commit flow
+    ref = "HEAD"
+    author = repo.default_signature
+    # Make an empty tree
+    tree_id = repo.TreeBuilder().write()
+    message = "Initialize scrapyard"
+    parents: object = []
+    commit = repo.create_commit(ref, author, author, message, tree_id, parents)
+    assert commit == repo.branches[branch_name].target
+    assert commit == repo.head.target
+
+
+def yard_init_command(args: argparse.Namespace) -> None:
+    yard_init(args.yard)
+
+
+def yard_commit(path: str, scrap_name: str, obj: Object) -> Tuple[str, str]:
+    git = _ensure_pygit2()
+    # Find the scrapyard
+    repo_path = git.discover_repository(path)
+    if repo_path is None:
+        raise ScrapError(f"Please create a scrapyard; {path!r} is not initialized")
+    repo = git.Repository(repo_path, git.GIT_REPOSITORY_OPEN_BARE)
+    serialized = serialize(obj)
+    # Make a git tree starting from previous commit's tree
+    prev_commit = repo.get(repo.head.target)
+    root = repo.TreeBuilder(prev_commit.tree)
+    obj_id = repo.create_blob(serialized)
+    # TODO(max): Figure out how to handle names like a/b; make directories?
+    root.insert(scrap_name, obj_id, git.GIT_FILEMODE_BLOB)
+    tree_id = root.write()
+    # Commit the tree
+    ref = repo.head.name
+    author = repo.default_signature
+    message = f"Update {scrap_name}"
+    parents = [repo.head.target]
+    commit = repo.create_commit(ref, author, author, message, tree_id, parents)
+    assert commit == repo.branches["trunk"].target
+    assert commit == repo.head.target
+    return scrap_name, obj_id
+
+
+def yard_commit_command(args: argparse.Namespace) -> None:
```

Comment:
@gregorybchris had an interesting idea that I'm sure he can describe more and better about local graph references getting linked up at commit time so you can have multiple files locally without committing each one every time

Reply:
I wrote some programs in scrapscript. It was genuinely a pretty good time all in all.

However, I wanted more modularity than the language currently allows. I felt a desire to group things into larger structures - modules in the general sense of the term (not a statement on within a file vs between files).

Two reasons:
1. Sometimes it's intuitive to think in terms of `let x in y` and other times it's more intuitive to think in `y where x`. Using modules that reference each other allows you to choose how you structure dependencies in your program. This is a higher level version of the argument for a pipe operator. Sometimes it's more intuitive to think about data flowing through chained operations rather than nested function applications. Some here, except instead you're writing module A before module B where B depends on A.
2. Many (most?) scraps will be small little programs. But some will be larger. The developer experience of having more than ~30 lines of scrapscript declines when you don't have a way to chunk your code into logical units. Sure, you could create records of functions, but the ergonomics of that aren't ideal.

The idea is to have some way to reference other scraps on your machine with a similar syntax to the scrapyard import syntax. Instead of using the scrap `chris/add-one` from a scrapyard, I could use `local/add-one` which somehow references a scrap defined in the current file or another file on disk.

Importantly this loose dependency graph need not leak into the scrapyard of well-defiend merkle trees. If I commit `chris/fun-program` which depends on `local/add-one` the scrapyard implementation can contract the graph of local scraps to register one fully-defined `chris/fun-program` that contains the full implementation of all dependencies.

The benefit is purely for the local development experience for large-ish scraps.

Reply:
Apologies, I'm incapable of writing short messages

Reply:
Super useful! Thanks for the exploration. I've done the same experiments and found some similar pain points.

If I understand you correctly, records are my preferred way of solving this problem:

```
mod . mod =
{ add = a -> b -> a + b
, sub = a -> b -> a - b
, add1 = mod::add 1
}
```

Or as it stands right now, `mod@add 1` :)